### DISCUSSION TITLE
get midi from another board (esp or arduino)

### DISCUSSION THREAD

**Discussion by fjgaston** [marcel-licence/esp32_midi_sampler#53] [Q&A]
Hi,
I don't know to much about midi signal but is there an easy way to send midi notes from an esp or arduino to the "esp32 midi sampler" ?
By the way, I'm trying to make it work with lyraT board, with 8388 codec, do you think of anything that could be wrong with this board ?

**Comment by souravlal** [comment]
Hi, you can use an Arduino to send midi messages. https://www.digikey.com/en/maker/projects/agm-summer-break-edition-raspberry-pi-midi-controller-synthesizer/45f94d1c75644494aaf7dda60e9b8693
Above is a good tutorial how to send midi signals using Arduino with button matrix. It uses an Arduino pro micro to send USB midi messages, however if you want to send serial midi in case for controlling ESP32 Audio-kit you can use the same code and circuit and just connect TX pin of the Arduino to send midi messages to ESP32 Audio-kit, in this case you can use any Arduino( nano, uno, etc...)

But if you don't want to use button matrix, you can use Arduino IO pins directly and if you want to extend the number of pins you can use multiplexers. Use this tutorial if you want to do this:-
https://www.notesandvolts.com/2016/07/arduino-midi-controller-multiplexers.html?m=1
In this project only Arduino Uno will work (Atmega328p IC will also work) if you are using multiple number of multiplexers.

You can use any optocoupler for Isolation, I would recommend digital optocouplers like 6N137 or H11L1 which doesn't miss out on any MIDI events.

**Reply by marcel-licence** [reply]
For first testing I can also recommend to use the NoteOn code in the setup function. In this case you can test the audio part independent from MIDI. Using the lyraT should be possible.

**Reply by fjgaston** [reply]
> For first testing I can also recommend to use the NoteOn code in the setup function. In this case you can test the audio part independent from MIDI. Using the lyraT should be possible.

How do I use the NoteOn code, is it this part of the code ?
`#if 1
    Sampler_NoteOn(0, 64, 1);
#endif`

I guess "if 1" means that it's activated ?
I loaded some organ samples in a samples directory in the SD card, should I get some sound ?
(The code compiles and upload without any error)

**Reply by marcel-licence** [reply]
https://youtu.be/Accf54fJn6U as shown in the video, there is are two function calls required: PatchManager_SetDestination(1, 1) to set the SD card as sample source.
Then Sampler_LoadPatchFile("/samples/yourFile.wav") to get this loaded.
There are also MIDI cc mapped to allow switching the destination (SD, littleFS) and also loading new samples aswell.

**Reply by fjgaston** [reply]
I finally uploaded the pet_bottle files with little FS and some sound comes out of my headphones.
Then for the sd card I use :
`//#define AUTO_LOAD_PATCHES_FROM_LITTLEFS
#define AUTO_LOAD_PATCHES_FROM_SD_MMC`

then :
`#ifdef AUTO_LOAD_PATCHES_FROM_SD_MMC
    PatchManager_SetDestination(1, 1);
    Sampler_LoadPatchFile("/samples/NoteTest.wav");
    //Sampler_LoadPatchFile("/samples/C2.wav");
#endif`

but I get no sound and the monitor outuputs :

`Free PSRAM: 0
[?25l[0;0HLoaded sample: 301              
--------------------------------
inL    [0;32m        [0m  outL   [0;32m        [0m
inR    [0;32m        [0m  outR   [0;32m        [0m
--------------------------------
301 of 2075068 bytes used       
O_______________________________
                                SD_MMC: /samples/NoteTest.wav`
but my audio file is 430 kbytes not 301 bytes

**Reply by marcel-licence** [reply]
Okease check if your NoteTest.wav is 16 bit, mono. I would recommend to export / save it with audacity. Some other programs add ID3 Tag information which can lead into errors.

**Reply by fjgaston** [reply]
Sorry for being a noob and thank you for all the support.
I experimented a few things and it seems that I need a ".bin" file with the same name of my audio file (I guess it sets up how the sample loop start and end) how do I set up this bin file ?
I managed to load another sample with littleFS and duplicate and rename "pet_bottle.bin" but I still didn't figure how to switch from one sample to another.

Also the problem with LyraT board is that only 2 buttons are available and 4 other buttons are touch buttons so I'll need to go deeper into the code to activate them.

**Reply by marcel-licence** [reply]
The bin file is optional. It will be generated when saving a sample. It stores the loop information and adsr settings. If you load a wav without bin the sample will be played without loop as one shot sample. You can also use the statements shown in my last video if you know the loop points by for example using audacity. 

If you have an IO free you could also use this as an analog input to use one input for multiple buttons if you have some restors left

**Reply by fjgaston** [reply]
I don't know if I did something wrong but at the moment the only way that I have to hear a sample is by adding a ".bin" file (and the loop is very short but it works) but if there is no bin file, there is no sound

**Reply by fjgaston** [reply]
I now tried 
`PatchManager_SetDestination(1, 1);
  Sampler_LoadPatchFile("/samples/1a.wav");
  Sampler_NoteOn(1,8,1);
  Sampler_NoteOff(1,8);
  sampleRecordCount--;
  Sampler_AddSection(64, 0, 213444, 6644, 140912, "1a");`

But it would not compile, with the error :
'sampleRecordCount' was not declared in this scope

**Reply by marcel-licence** [reply]
add this: extern `uint32_t sampleRecordCount;` at the top of the file. I will commit the update

**Reply by fjgaston** [reply]
awesome it works now !

**Reply by souravlal** [reply]
> The bin file is optional. It will be generated when saving a sample. It stores the loop information and adsr settings. If you load a wav without bin the sample will be played without loop as one shot sample. You can also use the statements shown in my last video if you know the loop points by for example using audacity.
> 
> If you have an IO free you could also use this as an analog input to use one input for multiple buttons if you have some restors left

@marcel-licence How can I save a sample? where will it be stored? Can I save a sample which I load from SD Card?

**Reply by marcel-licence** [reply]
Please look into z_config.ino. There is a function `Sampler_SavePatch` mapped to save your current settings -> wav and ino will be created

**Reply by souravlal** [reply]
@marcel-licence Where will it be saved? I mean which location?
Can it be done using buttons and display?
